datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,232 @@
+import os
+from typing import Dict, List, Optional
+from datastore.datastore import DataStore
+from models.models import (
+    DocumentChunk,
+    DocumentChunkMetadata,
+    DocumentMetadataFilter,
+    DocumentChunkWithScore,
+    DocumentMetadataFilter,
+    QueryResult,
+    QueryWithEmbedding,
+)
+import json
+import psycopg
+from psycopg import sql
+
+from services.date import to_unix_timestamp
+
+# Read environment variables for Postgres
+
+POSTGRES_HOST = os.environ.get("POSTGRES_HOST", "localhost")
+POSTGRES_PORT = int(os.environ.get("POSTGRES_PORT", 5433))
+POSTGRES_USERNAME = os.environ.get("POSTGRES_USERNAME", "postgres")
+POSTGRES_PASSWORD = os.environ.get("POSTGRES_PASSWORD", "")
+POSTGRES_DATABASE = os.environ.get("POSTGRES_DATABASE", "pgml_development")
+POSTGRES_TABLENAME = os.environ.get("POSTGRES_TABLENAME", "chatgpt_datastore")
+POSTGRES_SYNCHRONOUS_COMMIT = os.environ.get(
+    "POSTGRES_SYNCHRONOUS_COMMIT", "off"
+).lower()
+POSTGRES_UPSERT_COMMAND = os.environ.get("POSTGRES_UPSERT_COMMAND","COPY")
+
+assert POSTGRES_SYNCHRONOUS_COMMIT in ["on", "off"]
+assert POSTGRES_UPSERT_COMMAND in ["COPY","INSERT"]
+
+# OpenAI Ada Embeddings Dimension
+VECTOR_DIMENSION = 1536
+
+
+class PostgresDataStore(DataStore):
+    def __init__(self) -> None:
+        self.conn_params = {
+            "dbname": POSTGRES_DATABASE,
+            "user": POSTGRES_USERNAME,
+            "password": POSTGRES_PASSWORD,
+            "host": POSTGRES_HOST,
+            "port": POSTGRES_PORT,
+        }
+        conn = psycopg.connect(**self.conn_params)
+        # Insert the vector and text into the database
+        create_table = (
+            "CREATE TABLE IF NOT EXISTS %s (doc_id TEXT, chunk_id TEXT, text TEXT, embedding vector(%s), metadata JSONB)"
+            % (POSTGRES_TABLENAME, str(VECTOR_DIMENSION))
+        )
+        cur = conn.cursor()
+        cur.execute(create_table)
+        # Commit the transaction
+        conn.commit()
+        # Close the cursor and the connection
+        cur.close()
+        conn.close()
+
+    async def _upsert(self, chunks: Dict[str, List[DocumentChunk]]) -> List[str]:
```

Comment:
Returning the list of keys in the chunks seems ... redundant to me, but if that's how they do it...

Reply:
It returns only doc_ids and not chunk_ids.